File: lib/l10n/sync_languages.dart
import 'dart:convert';
import 'dart:io';

void main() async {
  final l10nDir = Directory('lib/l10n');
  final arabicFile = File('${l10nDir.path}/app_ar.arb');
  
  if (!await arabicFile.exists()) {
    print('Error: Arabic file not found at ${arabicFile.path}');
    return;
  }

  try {
    // Read Arabic file as reference
    final arabicContent = await arabicFile.readAsString();
    final arabicJson = json.decode(arabicContent) as Map<String, dynamic>;
    
    // Get all keys from Arabic file (excluding metadata)
    final arabicKeys = arabicJson.entries
        .where((entry) => !entry.key.startsWith('@@'))
        .toList();

    print('Found ${arabicKeys.length} keys in Arabic file');

    // Process each language file
    await for (var entity in l10nDir.list()) {
      if (entity is File && 
          entity.path.endsWith('.arb') && 
          !entity.path.endsWith('_ar.arb')) {
        await _syncFile(entity, arabicKeys);
      }
    }
  } catch (e) {
    print('Error: $e');
  }
}

Future<void> _syncFile(File file, List<MapEntry<String, dynamic>> arabicKeys) async {
  try {
    final content = await file.readAsString();
    final jsonData = json.decode(content) as Map<String, dynamic>;
    
    // Keep metadata
    final metadata = Map.fromEntries(
      jsonData.entries.where((entry) => entry.key.startsWith('@@')),
    );
    
    // Keep existing translations
    final existingTranslations = Map.fromEntries(
      jsonData.entries.where((entry) => !entry.key.startsWith('@@')),
    );
    
    // Create new data with all Arabic keys
    final newData = <String, dynamic>{};
    
    // Add metadata first
    newData.addAll(metadata);
    
    // Add all Arabic keys, keeping existing translations where they exist
    for (var entry in arabicKeys) {
      newData[entry.key] = existingTranslations[entry.key] ?? '';
    }
    
    // Sort keys for better readability
    final sortedKeys = newData.keys.toList()..sort();
    final sortedData = {
      for (var key in sortedKeys) key: newData[key],
    };
    
    // Write back to file
    final encoder = JsonEncoder.withIndent('  ');
    await file.writeAsString(encoder.convert(sortedData));
    
    print('Updated ${file.path} with ${arabicKeys.length} keys');
  } catch (e) {
    print('Error processing ${file.path}: $e');
  }
}

File: lib/l10n/sync_languages.py
import json
import os

def load_arb_file(file_path):
    with open(file_path, 'r', encoding='utf-8') as f:
        return json.load(f)

def save_arb_file(file_path, data):
    with open(file_path, 'w', encoding='utf-8') as f:
        json.dump(data, f, ensure_ascii=False, indent=2, sort_keys=True)

def get_all_keys(directory):
    all_keys = set()
    for filename in os.listdir(directory):
        if filename.startswith('app_') and filename.endswith('.arb'):
            try:
                data = load_arb_file(os.path.join(directory, filename))
                all_keys.update(key for key in data.keys() if not key.startswith('@@'))
            except Exception as e:
                print(f"Error processing {filename}: {e}")
    return sorted(all_keys)

def sync_language_files(directory):
    # Get all unique keys from all files
    all_keys = get_all_keys(directory)
    
    # Process each language file
    for filename in os.listdir(directory):
        if filename.startswith('app_') and filename.endswith('.arb'):
            filepath = os.path.join(directory, filename)
            try:
                # Load existing translations
                data = load_arb_file(filepath)
                
                # Create new data with all keys, preserving existing translations
                new_data = {}
                metadata = {k: v for k, v in data.items() if k.startswith('@@')}
                translations = {k: v for k, v in data.items() if not k.startswith('@@')}
                
                # Add metadata first
                new_data.update(metadata)
                
                # Add all keys with existing translations or empty strings
                for key in all_keys:
                    if key in translations:
                        new_data[key] = translations[key]
                    else:
                        new_data[key] = ""
                
                # Save the synchronized file
                save_arb_file(filepath, new_data)
                print(f"Updated {filename} with {len(all_keys)} keys")
                
            except Exception as e:
                print(f"Error updating {filename}: {e}")

if __name__ == "__main__":
    l10n_dir = os.path.dirname(os.path.abspath(__file__))
    sync_language_files(l10n_dir)
